tdrs-backend/tdpservice/parsers/parse.py

```diff
@@ -65,97 +51,170 @@ def parse_datafile(datafile):
     return errors
 
 
+def bulk_create_records(unsaved_records, line_number, header_count, batch_size=10000, flush=False):
+    """Bulk create passed in records."""
+    if (line_number % batch_size == 0 and header_count > 0) or flush:
+        try:
+            num_created = 0
+            num_expected = 0
+            for model, records in unsaved_records.items():
+                num_expected += len(records)
+                num_created += len(model.objects.bulk_create(records))
+            return num_created == num_expected, {}
+        except DatabaseError as e:
+            logger.error(f"Encountered error while creating datafile records: {e}")
+            return False, unsaved_records
+    return True, unsaved_records
+
+def bulk_create_errors(unsaved_parser_errors):
+    """Bulk create all ParserErrors."""
+    if unsaved_parser_errors:
+        ParserError.objects.bulk_create(list(itertools.chain.from_iterable(unsaved_parser_errors.values())))
+
+def evaluate_trailer(datafile, trailer_count, multiple_trailer_errors, is_last_line, line, line_number):
+    """Validate datafile trailer and return associated errors if any."""
+    if trailer_count > 1 and not multiple_trailer_errors:
+        return (True, [util.make_generate_parser_error(datafile, line_number)(
+                schema=None,
+                error_category=ParserErrorCategoryChoices.PRE_CHECK,
+                error_message="Multiple trailers found.",
+                record=None,
+                field=None
+                )])
+    if trailer_count == 1 or is_last_line:
+        record, trailer_is_valid, trailer_errors = schema_defs.trailer.parse_and_validate(
+            line,
+            util.make_generate_parser_error(datafile, line_number)
+        )
+        return (multiple_trailer_errors, None if not trailer_errors else trailer_errors)
+    return (False, None)
+
 def parse_datafile_lines(datafile, program_type, section):
     """Parse lines with appropriate schema and return errors."""
-    errors = {}
     rawfile = datafile.file
+    errors = {}
 
-    rawfile.seek(0)
     line_number = 0
-    schema_options = get_schema_options(program_type)
+    schema_manager_options = get_schema_manager_options(program_type)
+
+    unsaved_records = {}
+    unsaved_parser_errors = {}
+
+    header_count = 0
+    trailer_count = 0
+    prev_sum = 0
+    multiple_trailer_errors = False
 
+    # Note: it is unnecessary to call rawfile.seek(0) again because the generator
+    # automatically starts back at the begining of the file.
+    begin_transaction(transaction)
```

Comment:
I have two problems with "begin_transaction":
- I am very much leaned to wrapping it in try/except and turn transaction back in exception
- I am not clear what is the impact on other web tranactions while the autocommit is turned off. For example,  what if another user submits a file while autocommit is off? Is autocommit turned off for that transaction too?

Reply:
- We could certainly wrap the entirety of `parse_datafile_lines` in a try/except to avoid some random exception that might crop up. 
- As for the impact of transactions. From what I understand in the docs, transactions are scoped per process and per DB connection. So if another datafile is submitted and another worker/process executes it then there is no nesting/collision of transactions. If the same process is executing multiple calls to `parse_datafile_lines` and there arent unique DB connections, we will end up with nested transactions. Nested transactions are well supported however, if the first transaction is rolled back (i.e. the first datafile had some error where we deem we need to rollback) then any nested transaction would be rolled back to (this would include the second datafile).

Reply:
There arent specific docs surrounding that statement. It is just what I gathered from [this](https://docs.djangoproject.com/en/4.2/topics/db/transactions/ ) and [this](https://www.postgresql.org/docs/current/tutorial-transactions.html).